src/enzyme_ad/jax/Implementations/WhileLoopInfo.cpp

```diff
@@ -47,43 +49,89 @@ LogicalResult WhileLoopInfo::computeInfo() {
   step = inc.getOperand(1);
   start = op->getOperand(induct.getArgNumber());
   limit = cond.getOperand(1);
+  inclusive = direction == stablehlo::ComparisonDirection::LE;
 
   return success();
 }
 
-std::optional<int64_t> WhileLoopInfo::getConstantStep() {
-  DenseIntElementsAttr stepAttr;
+std::optional<DenseIntOrFPElementsAttr> WhileLoopInfo::getConstantStep() {
+  DenseIntOrFPElementsAttr stepAttr;
   if (!matchPattern(step, m_Constant(&stepAttr)))
     return std::nullopt;
-  return (*stepAttr.begin()).getSExtValue();
+  return stepAttr;
 }
 
-std::optional<int64_t> WhileLoopInfo::getConstantStart() {
-  DenseIntElementsAttr startAttr;
+std::optional<DenseIntOrFPElementsAttr> WhileLoopInfo::getConstantStart() {
+  DenseIntOrFPElementsAttr startAttr;
   if (!matchPattern(start, m_Constant(&startAttr)))
     return std::nullopt;
-  return (*startAttr.begin()).getSExtValue();
+  return startAttr;
 }
 
-std::optional<int64_t> WhileLoopInfo::getConstantLimit() {
-  DenseIntElementsAttr limitAttr;
+std::optional<DenseIntOrFPElementsAttr> WhileLoopInfo::getConstantLimit() {
+  DenseIntOrFPElementsAttr limitAttr;
   if (!matchPattern(limit, m_Constant(&limitAttr)))
     return std::nullopt;
-  return (*limitAttr.begin()).getSExtValue();
+  return limitAttr;
+}
+
+std::optional<int64_t> WhileLoopInfo::getConstantNumIters() {
+  DenseIntOrFPElementsAttr start = getConstantStart().value(),
+                           limit = getConstantLimit().value(),
+                           step = getConstantStep().value();
+
+  if (isa<mlir::FloatType>(start.getElementType())) {
+    auto start_f = *start.getValues<APFloat>().begin(),
+         limit_f = *limit.getValues<APFloat>().begin(),
+         step_f = *step.getValues<APFloat>().begin();
+
+    auto numIters_f = (limit_f - start_f) / step_f;
+
+    APSInt numIters_i(64, true);
+    bool isExact = false;
+    if (numIters_f.convertToInteger(numIters_i, llvm::APFloat::rmTowardZero,
+                                    &isExact) != llvm::APFloat::opOK ||
+        !isExact)
+      return std::nullopt;
+
+    return numIters_i.getSExtValue() + inclusive;
```

Comment:
I don't think this is mathematically correct if step != 1

e.g. 2 * i < 3    can be 0 or 1,  as well as 2 * i <= 3

Reply:
I think the better way to do the change is to add 1 to the limit, then use the < code